optuna/_optimize.py

```diff
@@ -0,0 +1,248 @@
+import copy
+import datetime
+import gc
+import math
+from typing import Callable
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Type
+import warnings
+
+import joblib
+from joblib import delayed
+from joblib import Parallel
+
+import optuna
+from optuna import exceptions
+from optuna import logging
+from optuna import progress_bar as pbar_module
+from optuna import storages
+from optuna import trial as trial_module
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+
+def _optimize(
+    study: "optuna.Study",
+    func: "optuna.study.ObjectiveFuncType",
+    n_trials: Optional[int] = None,
+    timeout: Optional[float] = None,
+    n_jobs: int = 1,
+    catch: Tuple[Type[Exception], ...] = (),
+    callbacks: Optional[List[Callable[["optuna.Study", FrozenTrial], None]]] = None,
+    gc_after_trial: bool = False,
+    show_progress_bar: bool = False,
+) -> None:
+    if not isinstance(catch, tuple):
+        raise TypeError(
+            "The catch argument is of type '{}' but must be a tuple.".format(type(catch).__name__)
+        )
+
+    if not study._optimize_lock.acquire(False):
+        raise RuntimeError("Nested invocation of `Study.optimize` method isn't allowed.")
+
+    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
+    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
```

Comment:
The progress bar doesn't have to be an attribute/increase the complexity of the `Study`, but can be a local object.